LifeCycleAssessment_Engine/Compute/EvaluateEnvironmentalProductDeclarationByMass.cs

```diff
@@ -47,44 +49,43 @@ public static partial class Compute
         private static EnvironmentalMetricResult EvaluateEnvironmentalProductDeclarationByMass(IElementM elementM, List<LifeCycleAssessmentPhases> phases, EnvironmentalProductDeclarationField field = EnvironmentalProductDeclarationField.GlobalWarmingPotential, bool exactMatch = false)
         {
             double volume = elementM.ISolidVolume();
-            List<double> epdVal = elementM.GetEvaluationValue(field, phases, QuantityType.Mass, exactMatch);
             List<double> gwpByMaterial = new List<double>();
-            List<double> volumeByRatio = elementM.IMaterialComposition().Ratios.Select(x => volume * x).ToList();
-            List<double> densityOfMassEpd = Query.GetEPDDensity(elementM);
-            List<double> massOfObj = new List<double>();
+            MaterialComposition mc = elementM.IMaterialComposition();
+            List<Material> matList = mc.Materials.ToList();
 
-            if (densityOfMassEpd == null)
+            List<double> epdVals = elementM.GetEvaluationValue(field, phases, QuantityType.Mass, exactMatch);
+            if (epdVals == null || epdVals.Where(x => !double.IsNaN(x)).Sum() <= 0)
             {
-                BH.Engine.Reflection.Compute.RecordError("Density could not be found. Material density is required for all objects using Mass-based evaluations.");
+                BH.Engine.Reflection.Compute.RecordError($"No value for {field} can be found within the supplied EPD.");
                 return null;
             }
 
-            for (int x = 0; x < volumeByRatio.Count; x++)
-            {
-                massOfObj.Add(volumeByRatio[x] * densityOfMassEpd[x]);
-            }
-
-            for (int x = 0; x < epdVal.Count; x++)
+            for (int i = 0; i < matList.Count(); i++)
             {
-                if (double.IsNaN(epdVal[x]))
-                    gwpByMaterial.Add(double.NaN);
-                else
+                double density;
+                List<EnvironmentalProductDeclaration> materialEPDs = matList[i].Properties.OfType<EnvironmentalProductDeclaration>().ToList();
+                if (materialEPDs.Any(x => x.QuantityType == QuantityType.Mass))
                 {
-                    gwpByMaterial.Add(epdVal[x] * massOfObj[x]);
-                }
-            }
+                    double volumeOfMaterial = mc.Ratios[i] * volume;
+                    List<double> densityOfMassEpd = materialEPDs[0].GetEPDDensity();
+                    if (densityOfMassEpd == null || densityOfMassEpd.Count() == 0)
+                    {
+                        BH.Engine.Reflection.Compute.RecordError("Density could not be found. Material density is required for all objects using Mass-based evaluations.");
```

Comment:
Looking good! Let's prompt the user to add a DensityFragment for all objects using Mass-based QuantityType EPDs.

Reply:
Thanks, the error message has been updated